Minimal enterprise auth manager

henriknyman opened the pull request:
A minimal enterprise auth manager based on Apache Shiro,
with a custom file user realm mapping to our flat auth file
plus a new custom flat file mapping roles to multiple users,
and a static mapping between predefined roles and permissions.

henriknyman committed:
Override AuthManager using Apache Shiro

henriknyman committed:
Add authorization using static groups

- Add a single group per user to the file user repository
- Authorize using a static mapping from group to role
- Add user access mode tests
- Add a separate shiro auth manager test file

henriknyman committed:
Move ShiroAuthManager to enterprise

- Define constants for the fixed role name strings

henriknyman committed:
Refactor password policy

- Created PasswordPolicy interface
- Extracted password validation from BasicAuthSubject

henriknyman committed:
Remove legacy authenticate() method

henriknyman committed:
Add cache manager to enterprise auth

- Currently the Shiro supported EhCacheManager
- Fix licenses

henriknyman committed:
Refactor password policy

- Set password policy for ShiroAuthManager
- Simplify UserAccount to just use SimpleAuthorizationInfo

henriknyman committed:
Simplify password policy

Setting same password is a fundamental failure and does not need to have a
duplicate check in the password policy.

henriknyman committed:
Fix licenses

OliviaYtterbrink commented on community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java:
```diff
@@ -173,7 +169,7 @@ public void setUserPassword( String username, String password ) throws IOExcepti
 
         if ( existingUser.credentials().matchesPassword( password ) )
```
Check password policy here instead of in setPassword?

henriknyman replied:
The password policy functionality is duplicated in REST UserService, with its own expected status code, hence the division of a setUserPassword (it is the old API renamed)
Here is an attempt to remove password policy from UserService and move validation into setUserPassword (but it changes the expected status code of one test..): https://github.com/henriknyman/neo4j/commit/4d79ced5def19fd9ea9ac599d050b371b2fa3f32

craigtaverner commented on community/security/src/main/java/org/neo4j/server/security/auth/User.java:
```diff
@@ -94,6 +104,7 @@ public boolean equals( Object o )
     public int hashCode()
     {
         int result = name != null ? name.hashCode() : 0;
+        result = 31 * result + ( group != null ? group.hashCode() : 0);
```
Why would the group be part of the hashCode? That seems to imply that changing a users group somehow changes the identity of the user?

craigtaverner replied:
Or is this only used to decide if the line in the auth file should be updated and the current location of a group field in that file means we need to do this?

henriknyman replied:
Equivalence of User _objects_ is only used to detect concurrent modification of the auth repository  (throwing `ConcurrentModificationException`), so this `hashCode` just matches `equals`.
Whenever the user identity is compared it is by `user.name()`.

craigtaverner replied:
If we keep groups outside auth, then concurrent modification and identity would be the same.

craigtaverner commented on community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java:
```diff
@@ -72,21 +72,28 @@
     private String serialize( User user )
     {
         return join( userSeparator, user.name(),
+                user.group(),
```
If group was previously not part of the auth file, perhaps we should not add it, but add the groups file? That would remove a potentially unclean intermediate state.

henriknyman replied:
Last time we discussed this you agreed that it would make sense to keep the Unixy way of a default group in the user file for convenience, and then add a group file. I am not sure what you mean by unclean intermediate state?

craigtaverner commented on community/security/src/test/java/org/neo4j/server/security/auth/BasicAuthManagerTest.java:
```diff
@@ -59,18 +59,18 @@ public void shouldFindAndAuthenticateUserSuccessfully() throws Throwable
         // Given
         final InMemoryUserRepository users = new InMemoryUserRepository();
-        final User user = new User( "jake", Credential.forPassword( "abc123" ), false );
+            final User user = new User( "jake", "admin", Credential.forPassword( "abc123" ), false );
```
I'm more comfortable having the tests also reflect the design of creating users and groups separately and assigning users to groups, instead of a group as a field of a user.

craigtaverner replied:
Since this is the community auth manager, we could also argue that there is no concept of groups here.

henriknyman replied:
We are no actually testing the group, it is just passing something in.

We already have a plan to add multiple groups as a separate card. This is just the minimal first iteration.

craigtaverner commented on community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java:
```diff
@@ -77,7 +83,7 @@ public void shouldStoreAndRetriveUsersByName() throws Exception
         // Given
         FileUserRepository users = new FileUserRepository( authFile, NullLogProvider.getInstance() );
-        User user = new User( "jake", Credential.INACCESSIBLE, true );
+        User user = new User( "jake", "admin", Credential.INACCESSIBLE, true );
```
As above, I feel like in community there is no concept of groups. Should we be testing for this here?

henriknyman replied:
We are not testing the group. The reason group is in community is for backwards compatibility of the auth file and to simplify the architecture. The group is ignored and has no meaning to the community auth manager.

craigtaverner commented on enterprise/NOTICE.txt:
```diff
@@ -1,29 +1,26 @@
-LICENSE Version 3 (http://www.fsf.org/licensing/licenses/agpl-3.0.html)
-("AGPL") to all third parties and that license, as required by the AGPL,
-is included in the LICENSE.txt file.
+The software ("Software") is developed and owned by Network Engine
```
Should this be part of a security PR? Perhaps it should be at least in a separate commit?

henriknyman replied:
Well, this was what the licensing plugin changed, and otherwise we did not pass the Team City tests. I guess there was a recent update to the licence? It is a bit weird why we would have to change this though... 

It is in a separate commit.

henriknyman committed:
Move password policy from UserService

- Share one password policy for both REST and Bolt
- Simplify password policy to not require AuthSubject

craigtaverner commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedGroupsAccountBuilder.java:
```diff
@@ -0,0 +1,53 @@
+        case "admin":
+        case "architect":
+        case BasicAuthManager.DEFAULT_GROUP:
+            return Collections.singleton( ShiroAuthSubject.SCHEMA_READ_WRITE );
```
Here we seems to be using the term 'Role' to mean Permission. I thought 'Role' was 'Group'. We should probably clarify and correct to match the current designs.

henriknyman replied:
Yes this pre-dates the latest design documents. I had a sketch of how to map the resource permission terminology we drew up in the last design document to Shiro roles and wildcard permission, but was waiting for it to stabilize.

craigtaverner commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java:
```diff
@@ -0,0 +1,164 @@
+            IllegalCredentialsException
+    {
+        assertAuthEnabled();
+        return realm.newUser( username, group, initialPassword, requirePasswordChange );
```
I'm still uncomfortable with this hard linking of a user to a single group.

henriknyman replied:
We have a card to change this :)

henriknyman committed:
Support group/role repository in enterprise

- Flat group file mapping group to users
- Statically map roles to Shiro wildcard permissions

henriknyman committed:
Remove default group from user repository

henriknyman committed:
Rename group to role

henriknyman committed:
Change role store setting

- Add a new directory setting dbms.directories.auth that takes precedence
over the legacy unsupported.dbms.security.auth_store.location
- Make the file names of the auth and roles files an internal thing
- Use the directory of the legacy auth store location setting as a fallback

henriknyman committed:
Handle ConcurrentModificationException

- Cleanup exception usage
- Retry on ConcurrentModificationException
- Add RoleManager interface (some implementation still TODO)

henriknyman committed:
Make role management atomic

- Add com.googlecode.thread-weaver for concurrency tests to
deterministically test concurrent updates with a sequential flow
- Add atomicity tests for addUserToRole and deleteUser
- Share common functionality in AbstractRoleRepository
 (this also increases test coverage)

henriknyman committed:
Code style changes and cleanup

craigtaverner commented on community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java:
```diff
@@ -233,14 +239,18 @@ public void shouldIgnoreInvalidEntries() throws Throwable
         Files.createDirectories( authFile.getParent() );
         Files.write( authFile, UTF8.encode(
-                "neo4j:fc4c600b43ffe4d5857b4439c35df88f:SHA-256,A42E541F276CF17036DB7818F8B09B1C229AAD52A17F69F4029617F3A554640F,FB7E8AE08A6A7C741F678AD22217808F:\n" ) );
+                "neo4j:admin:fc4c600b43ffe4d5857b4439c35df88f:SHA-256," +
```
Is this a residual group field? Should be moved, I assume?

henriknyman replied:
This has slightly wrong name/assertions/error messages. Changed the test in a new PR to reflect the actual behaviour change (we now fail hard on a wrongly formatted file).

craigtaverner commented on community/server/src/test/java/org/neo4j/server/rest/dbms/UserServiceTest.java:
```diff
@@ -200,25 +226,22 @@ public void shouldReturn404WhenChangingPasswordIfDifferentUser() throws Exceptio
 
     @Test
-    public void shouldReturn404WhenChangingPasswordIfUnknownUser() throws Exception
+    public void shouldReturn422WhenChangingPasswordIfUnknownUser() throws Exception
```
I thought the idea was to standardize on less informative return codes?

henriknyman replied:
422 is what we are doing for the other types of illegal requests so you could argue it is less informative (no distinction in error code). We do however still provide exact descriptions in the messages.
I think standardising error message responses across REST and Bolt should be done as a separate PR.

craigtaverner commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseAuthManagerFactory.java:
```diff
@@ -43,9 +50,30 @@ public EnterpriseAuthManagerFactory()
+        else
+        {
+            // Fallback on the directory of the legacy setting
+            userStoreFile = config.get( GraphDatabaseSettings.auth_store );
```
Should we support legacy settings? Perhaps since this is new code, we should not. Or do we have an intermediate state where current configs have no default?

henriknyman replied:
Current config have no default, so we are using the existing default directory of the old setting. We need to revisit this after discussing with operability. Created a card for tracking it.
